### PR TITLE
i#3427 detach flakiness: Add two derivatives of the api.detach_spawn test.

### PR DESCRIPTION
**Pull request by hgreving2304**
The remaining items in #3427 were non-issues, see comment in #3427. Instead, this patch is
adding two tests. Both are derivatives of api.detach_spawn.

api.detach_spawn_quick_exit: Adds start/stop_and_cleanup re-attach to api.detach_spawn.
It only spawns one thread. The thread is synch'd such that it exits while running native.
This case is not supported by DynamoRIO's dr_app_stop w/o detach, but must work w/
dr_app_stop_and_cleanup().

api.detach_spawn_stress: Adds start/stop_and_cleanup re-attach to api.detach_spawn.
Increase the number of spawned parent and child threads to 25 each, spawning a total
number of 25 + 25 * 25 = 650 threads at a time, but limit the re-spawn loop. Like
api.detach_spawn, this test is currently flaky in 32-bit, and very rarely flaky in 64-bit.
#3427 covers the 64-bit failure. I have added a comment to #2694.

Issues: #3427, #2694

### CODE REVIEW

**Comment by hgreving2304** [comment]
I will adjust the copyright dates in the tests before merge.

**Comment by hgreving2304** [comment]
New api.detach_spawn_stress test already failed on Travis. Marking it FLAKY now even for 64-bit. Locally it's rare/hard to make it fail, not sure if same bug. Locally I am running the test in an infinite loop trying to get a core dump.

File: suite/tests/api/detach_spawn_stress.c [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    /* We setup and start at once to avoid process memory changing much between
+     * the two.
+     */
+    dr_app_setup_and_start();
+    wait_cond_var(parent_ready);
+    thread_sleep(50);
+
+    if (!dr_app_running_under_dynamorio())
+        print("ERROR: should be running under DynamoRio before calling "
+              "dr_app_stop_and_cleanup()\n");
```
Comment (derekbruening) [inline suite/tests/api/detach_spawn_stress.c]: Many of the same comments: style violation; DynamoRIO.

File: suite/tests/CMakeLists.txt [hunk 2]
```diff
@@ -2323,12 +2323,19 @@ if (CLIENT_INTERFACE)
     if (NOT WIN32) # XXX i#2611: fix for Windows
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Done.

File: suite/tests/CMakeLists.txt [hunk 3]
```diff
@@ -2327,8 +2327,14 @@ if (CLIENT_INTERFACE)
         # FIXME i#2694: failing sometimes on 32-bit Linux.
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Sure, done.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+#else
+#    define VPRINT(...) /* nothing */
+#endif
+
+THREAD_FUNC_RETURN_TYPE
+thread_func(void *arg)
+{
+    signal_cond_var(thread_ready);
+    while (thread_do_stuff)
+        ;
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Ok.

[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Probably would work but I wanted the thread actually do something and not yield every iteration such that it is more likely do actually exit the loop and exits once "do stuff" is false.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    print("Running under DynamoRIO\n");
+
+    dr_app_stop_and_cleanup();
+
+    thread_sleep(50);
+
+    if (dr_app_running_under_dynamorio())
+        print("ERROR: should not be running under DynamoRio before calling "
+              "dr_app_stop()\n");
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Ok, to be consistent, I added {} to the other running_under_dynamorio one liners as well.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ * DAMAGE.
+ */
+
+/* This test is a modified version of api.detach_spawn. It adds start/stop re-attach to
+ * api.detach_spawn. It only spawns one thread. The thread is synch'd such that it exits
+ * while running native. This case is not supported by DynamoRIO's dr_app_stop w/o detach,
+ * but must work w/ dr_app_stop_and_cleanup()
+ */
+
+#include <assert.h>
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Done

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+#include <stdio.h>
+#include <math.h>
+#include <stdint.h>
+#include "configure.h"
+#include "dr_api.h"
+#include "tools.h"
+#include "thread.h"
+#include "condvar.h"
+#include <sys/syscall.h>
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Yes, it was used to debug to read tid. The gettid call is unused and can get removed as well.

File: suite/tests/api/detach_spawn_stress.c [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This test is a modified version of api.detach_spawn. It adds start/stop re-attach to
+ * api.detach_spawn. It increases the number of spawned parent and child threads to 25
+ * each, spawning a total number of 25 + 25 * 25 = 650 threads at a time.
+ */
+
+#include <assert.h>
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_stress.c]: Done.

File: suite/tests/api/detach_spawn_stress.c (hunk 1 above)
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_stress.c]: Done.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+/* This test is a modified version of api.detach_spawn. It adds start/stop re-attach to
+ * api.detach_spawn. It only spawns one thread. The thread is synch'd such that it exits
+ * while running native. This case is not supported by DynamoRIO's dr_app_stop w/o detach,
+ * but must work w/ dr_app_stop_and_cleanup()
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <math.h>
+#include <stdint.h>
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Done.

File: suite/tests/api/detach_spawn_quick_exit.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/api/detach_spawn_quick_exit.c]: OK but the rule is about *lines* not *statements* as presumably you have seen in the style guide.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    /* We setup and start at once to avoid process memory changing much between
+     * the two.
+     */
+    wait_cond_var(thread_ready);
+
+    dr_app_setup_and_start();
+
+    if (!dr_app_running_under_dynamorio())
+        print("ERROR: should be running under DynamoRio before calling dr_app_stop()\n");
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Fixed.

File: suite/tests/api/detach_spawn_quick_exit.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Yep.

File: suite/tests/api/detach_spawn_quick_exit.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Done.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+    for (i = 0; i < NUM_THREADS; ++i)
+        threads[i] = create_thread(thread_func, NULL);
+
+    pid_t tid = dynamorio_syscall(SYS_gettid, 0);
+
+    /* We setup and start at once to avoid process memory changing much between
+     * the two.
+     */
+    wait_cond_var(thread_ready);
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: threads[NUM_THREADS] => thread.
NUM_THREADS removed.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    dr_app_setup_and_start();
+
+    if (!dr_app_running_under_dynamorio())
+        print("ERROR: should be running under DynamoRio before calling dr_app_stop()\n");
+
+    print("Running under DynamoRIO\n");
+
+    dr_app_stop_and_cleanup();
+
+    thread_sleep(50);
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Removed.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+#    define VPRINT(...) print(__VA_ARGS__)
+#else
+#    define VPRINT(...) /* nothing */
+#endif
+
+THREAD_FUNC_RETURN_TYPE
+thread_func(void *arg)
+{
+    signal_cond_var(thread_ready);
+    while (thread_do_stuff)
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Done.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+/* This test is a modified version of api.detach_spawn. It adds start/stop re-attach to
+ * api.detach_spawn. It only spawns one thread. The thread is synch'd such that it exits
+ * while running native. This case is not supported by DynamoRIO's dr_app_stop w/o detach,
+ * but must work w/ dr_app_stop_and_cleanup()
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <math.h>
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Removed in both tests.

**Comment by hgreving2304** [comment]
PTAL / could get merged.

File: suite/tests/api/detach_spawn_quick_exit.c [hunk 15]
```diff
@@ -63,8 +58,10 @@ THREAD_FUNC_RETURN_TYPE
 thread_func(void *arg)
 {
     signal_cond_var(thread_ready);
-    while (thread_do_stuff)
+    while (!thread_should_exit) {
+        /* Deliberately empty. */
         ;
```
Comment (derekbruening) [inline suite/tests/api/detach_spawn_quick_exit.c]: With the braces no need for the `;`

Comment (hgreving2304) [reply suite/tests/api/detach_spawn_quick_exit.c]: Done.

**Pull request merged**